zebra-state/src/service/read/difficulty.rs

```diff
@@ -18,33 +19,85 @@ use crate::{
             AdjustedDifficulty,
         },
         finalized_state::ZebraDb,
+        read::tree::history_tree,
         NonFinalizedState,
     },
-    GetBlockTemplateChainInfo,
+    BoxError, GetBlockTemplateChainInfo, HashOrHeight,
 };
 
+use crate::service::block_iter::Iter;
+
 /// Returns the [`GetBlockTemplateChainInfo`] for the current best chain.
 ///
 /// # Panics
 ///
-/// If we don't have enough blocks in the state.
-pub fn difficulty_and_time_info(
+/// - If we don't have enough blocks in the state.
+/// - If a consistency check fails `RETRIES` times.
+pub fn get_block_template_chain_info(
     non_finalized_state: &NonFinalizedState,
     db: &ZebraDb,
     tip: (Height, Hash),
```

Comment:
This `tip` argument comes from the state, so it needs to be inside the tip check in `relevant_chain_and_history_tree()`. Otherwise, it could be different from the tips in the tip check, and we wouldn't know.

The tip check already gets two tips, so you can just use one of those.

Reply:
Fixed in commit 0dfe898 by re-using the tip from the tip query. I also split the tip tuple into a separate height and hash argument. This makes it easier to write and review code. (I can never remember if `.0` is the height or the hash.)